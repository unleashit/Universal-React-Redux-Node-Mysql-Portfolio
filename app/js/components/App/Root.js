import React, { Component } from 'react';

class Root extends Component {
    renderInitialState() {
        if (this.props.initialState) {
            const innerHtml = `window.__INITIAL_STATE__ = ${JSON.stringify(
                this.props.initialState
            )}`;
            return <script dangerouslySetInnerHTML={{ __html: innerHtml }} />;
        }
    }

    renderEnvironment() {
        const innerHtml = `window.__ENVIRONMENT__ = '${__ENVIRONMENT__};'
            window.__GOOGLE_ANALYTICS__ = '${__GOOGLE_ANALYTICS__}';
            window.__LIVE_CHAT_ADMIN_NAME__ = '${__LIVE_CHAT_ADMIN_NAME__}';`;
        return <script dangerouslySetInnerHTML={{ __html: innerHtml }} />;
    }

    lazyLoadScripts() {
        const innerHtml = `
                function lazyLoadScript(scrpt) {
                    var element = document.createElement("script");
                    element.src = scrpt;
                    element.async = true;
                    document.body.appendChild(element);
                }`;
        return <script dangerouslySetInnerHTML={{ __html: innerHtml }} />;
    }

    lazyLoadScript(scrpt) {
        if (process.env.NODE_ENV !== 'production' && scrpt.global) {
            return <script src={scrpt.devSrc} />;
        } else {
            const innerHtml = `
                window.addEventListener('load', lazyLoadScript.bind(null, '${
                    scrpt.src
                }'), false);`;
            return <script dangerouslySetInnerHTML={{ __html: innerHtml }} />;
        }
    }

    render() {
        const head = this.props.head;
        const attrs = head.htmlAttributes.toComponent();

        return (
            <html {...attrs}>
                <head>
                    {head.title.toComponent()}
                    {head.meta.toComponent()}
                    <link
                        rel="apple-touch-icon"
                        sizes="180x180"
                        href="/images/favicons/apple-touch-icon.png"
                    />
                    <link
                        rel="icon"
                        type="image/png"
                        href="/images/favicons/favicon-32x32.png"
                        sizes="32x32"
                    />
                    <link
                        rel="icon"
                        type="image/png"
                        href="/images/favicons/android-chrome-192x192.png"
                        sizes="192x192"
                    />
                    <link
                        rel="icon"
                        type="image/png"
                        href="/images/favicons/favicon-16x16.png"
                        sizes="16x16"
                    />
                    <link
                        rel="manifest"
                        href="/images/favicons/manifest.json"
                    />
                    <link
                        rel="mask-icon"
                        href="/images/favicons/safari-pinned-tab.svg"
                        color="#5bbad5"
                    />
                    <link
                        rel="shortcut icon"
                        href="/images/favicons/favicon.ico"
                    />
                    <meta
                        name="google-site-verification"
                        content="PcZbHamu8N77ConcEm3oykmUEroNsJhpM6ZTrnvEO40"
                    />
                    <meta name="format-detection" content="telephone=no" />
                    <meta
                        name="msapplication-config"
                        content="/images/favicons/browserconfig.xml"
                    />
                    <meta name="theme-color" content="#ffffff" />
                    <meta name="referrer" content="origin" />
                    {head.link.toComponent()}
                    {this.renderEnvironment()}
                    {this.lazyLoadScripts()}
                </head>
                <body>
                    <div
                        id="root"
                        dangerouslySetInnerHTML={{ __html: this.props.content }}
                    />
                    {this.renderInitialState()}
                    {head.script.toComponent()}
                    {this.lazyLoadScript({
                        src: '/js/global.min.js',
                        devSrc: '/js/global.js',
                        global: true
                    })}
                    {this.lazyLoadScript({
                        src:
                            'https://cdnjs.cloudflare.com/ajax/libs/smooth-scroll/11.0.2/js/smooth-scroll.min.js'
                    })}
                </body>
            </html>
        );
    }
}

export default Root;
