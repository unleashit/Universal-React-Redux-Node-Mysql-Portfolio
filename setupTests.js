import React from 'react';
import Enzyme, { shallow, render, mount } from 'enzyme';
import fetchMock from 'fetch-mock';
import Adapter from 'enzyme-adapter-react-16';
import { Provider } from 'react-redux';
import combineReducers from './app/js/reducers';
import { createStore, applyMiddleware } from 'redux';
import thunkMiddleware from 'redux-thunk';
import configureStore from 'redux-mock-store';
import combinedReducers from './app/js/reducers';

//jasmine.DEFAULT_TIMEOUT_INTERVAL = 30000;

// React 16 Enzyme adapter
Enzyme.configure({ adapter: new Adapter() });

// Make Enzyme functions and React available in all test files without importing
global.React = React;
global.shallow = shallow;
global.render = render;
global.mount = mount;
global.fetchMock = fetchMock;

// global mocks
window.alert = global.alert = jest.fn((msg) => console.log(msg));

// redux helpers

export function wrapActualStore(
    Component,
    props = {},
    reducers = combinedReducers
) {
    const createStoreWithMiddleware = applyMiddleware(thunkMiddleware)(
        createStore
    );
    const store = createStoreWithMiddleware(reducers, {});

    return (
        <Provider store={store}>
            <Component {...props} />
        </Provider>
    );
}

export function wrapMockStore(Component, initialState = {}, props = {}) {
    const mockStore = configureStore([thunkMiddleware]);
    const mockedStore = mockStore(initialState);

    return (
        <Provider store={mockedStore}>
            <Component {...props} />
        </Provider>
    );
}

export function createMockStore(initialState = {}) {
    const mockStore = configureStore([thunkMiddleware]);
    return mockStore(initialState);
}
