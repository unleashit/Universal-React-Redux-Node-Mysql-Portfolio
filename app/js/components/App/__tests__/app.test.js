import { App } from '../App';
import Helmet from 'react-helmet';
import Footer from '../../footer/footer';
import LiveChatContainer from '../../ReactHelpDesk/client/liveChatContainer';
import LiveChatLauncher from '../../ReactHelpDesk/client/chatLauncher';
import { wrapActualStore } from '../../../../../setupTests';

describe('<App />', () => {
    let wrapper;
    const props = {
        liveChat: {
            chatOpen: false,
        },
        toggleChat: jest.fn(),
    };

    // make it output the production JS because js/css
    // filename changes in helmet
    process.env.NODE_ENV = 'production';

    beforeEach(() => {
        props.toggleChat.mockReset();
        wrapper = shallow(<App {...props} />);
    });

    afterAll(() => {
        process.env.NODE_ENV = 'test';
    });

    it('Renders', () => {
        expect(wrapper.find('.page-wrapper')).toHaveLength(1);
    });

    it('matches snapshot (without footer)', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('Renders a route', () => {
        const children = <div className="dummy-route" />;
        wrapper = shallow(<App {...props} children={children} />);
        expect(wrapper.find('.dummy-route')).toHaveLength(1);
    });

    it('React Helmet is used', () => {
        expect(wrapper.find(Helmet)).toHaveLength(1);
    });

    describe('Renders global components', () => {
        it('<LiveChatContainer />', () => {
            expect(wrapper.find(LiveChatContainer)).toHaveLength(1);
        });

        it('<LiveChatLauncher />', () => {
            expect(wrapper.find(LiveChatLauncher)).toHaveLength(1);
        });

        it('renders <Footer /> only when it should', () => {
            let children = {
                props: {
                    route: {
                        path: '/index',
                    },
                },
            };

            // footer is expected
            wrapper = shallow(<App {...props} children={children} />);
            expect(wrapper.find(Footer)).toHaveLength(1);

            // footer is not expected
            children.props.route.path = undefined;
            wrapper = shallow(<App {...props} children={children} />);
            expect(wrapper.find(Footer)).toHaveLength(0);

            children.props.route.path = '*';
            wrapper = shallow(<App {...props} children={children} />);
            expect(wrapper.find(Footer)).toHaveLength(0);
        });
    });

    it('toggleChatHandler() closes chat if open', () => {
        jest.spyOn(App.prototype, 'toggleChatHandler');
        props.liveChat.chatOpen = true;
        wrapper = shallow(<App {...props} />);
        wrapper.instance().toggleChatHandler();
        expect(props.toggleChat).toHaveBeenCalledTimes(1);
    });
});
