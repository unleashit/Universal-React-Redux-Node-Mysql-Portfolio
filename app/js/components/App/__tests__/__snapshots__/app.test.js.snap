// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> matches snapshot (without footer) 1`] = `
<div
  className="page-wrapper"
>
  <div
    className="content-wrapper"
    onClick={[Function]}
  >
    <HelmetWrapper
      defer={true}
      encodeSpecialCharacters={true}
      link={
        Array [
          Object {
            "href": "/css/global.min.css",
            "rel": "stylesheet",
          },
          Object {
            "async": true,
            "href": "https://fonts.googleapis.com/css?family=Oswald|Sanchez&display=block",
            "rel": "stylesheet",
          },
        ]
      }
      meta={
        Array [
          Object {
            "char-set": "utf-8",
          },
          Object {
            "content": "Full Stack Engineer in Berkeley. Focused on React, Angular, Node.Js, Drupal, UI/UX. San Francisco Bay Area, East Bay, San Francisco and Oakland.",
            "name": "description",
          },
          Object {
            "content": "width=device-width, initial-scale=1",
            "name": "viewport",
          },
        ]
      }
      title="Front End Engineer in Berkeley, CA specializing in React, Javascript, Node.Js, Angular and Drupal"
      titleTemplate="Jason Gallagher - %s"
    />
  </div>
  <Connect(LiveChatLauncher) />
  <Connect(LiveChatContainer) />
</div>
`;
